security/advancedtls/crl_provider.go

```diff
@@ -0,0 +1,229 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"time"
+)
+
+const defaultCRLRefreshDuration = 1 * time.Hour
+
+// CRLProvider is the interface to be implemented to enable custom CRL provider
+// behavior.
+//
+// The interface defines how the data is read, but doesn't prescribe a way
+// CRL are loaded and stored. Such implementations can be used in
+// RevocationConfig of advancedtls.ClientOptions and/or
+// advancedtls.ServerOptions .
+//
+// TODO(erm-g): Add link to related gRFC once it's ready.
+// Please refer to https://github.com/grpc/proposal/ for more details.
+type CRLProvider interface {
+	// CRL accepts x509 Cert and returns back related CRL struct. The CRL struct
+	// can be nil, can contain empty or non-empty list of revkoed certificates.
+	// Callers are expected to use the returned value as read-only.
+	CRL(cert *x509.Certificate) (*CRL, error)
+}
+
+// StaticCRLProvider implements CRLProvider interface by accepting CRL structs
+// and storing them in-memory.
+type StaticCRLProvider struct {
+	// TODO CRL is sort of our internal representation - provide an API for
+	// people to read into it, or provide a simpler type in the API then
+	// internally convert to this form
+	crls map[string]*CRL
+}
+
+// MakeStaticCRLProvider returns a new instance of the StaticCRLProvider.
+func MakeStaticCRLProvider() *StaticCRLProvider {
+	p := StaticCRLProvider{}
+	p.crls = make(map[string]*CRL)
+	return &p
+}
+
+// AddCRL adds/updates provided CRL to in-memory storage.
+func (p *StaticCRLProvider) AddCRL(crl *CRL) {
+	key := crl.CertList.Issuer.ToRDNSequence().String()
+	p.crls[key] = crl
+}
+
+// CRL returns CRL struct if it was previously loaded by calling AddCRL.
+func (p *StaticCRLProvider) CRL(cert *x509.Certificate) (*CRL, error) {
+	key := cert.Issuer.ToRDNSequence().String()
+	return p.crls[key], nil
+}
+
+// Options represents a data structure holding a
+// configuration for FileWatcherCRLProvider.
+type Options struct {
```

Comment:
I think we should name this `FileWatcherOptions`

Reply:
Alternatively (and this is a go style thing so I'm not 100% sure) we could just eschew the `Options` struct and have the `MakeFileWatcherCrlProvider` just take these arguments directly

Reply:
Done